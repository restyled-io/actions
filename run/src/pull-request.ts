/* Copyright (C) 2024 Patrick Brisbin
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
import * as fs from "fs";
import * as temp from "temp";

import * as core from "@actions/core";
import * as github from "@actions/github";

import { readProcess } from "./process";

type GitHubClient = ReturnType<typeof github.getOctokit>;

export type PullRequest = {
  number: number;
  title: string;
  headRef: string;
  headSha: string;
  restyleArgs: string[];
  restyleDiffBase: DiffBase;
  diff: string | null;
};

export type DiffBase = { tag: "unknown" } | { tag: "known"; sha: string };

export async function getPullRequest(
  client: GitHubClient,
  paths: string[],
): Promise<PullRequest> {
  const base = await getDiffBase();
  const pr = github.context.payload.pull_request;
  core.debug(`payload.pull_request: ${JSON.stringify(pr)}`);

  if (!pr) {
    return fakePullRequest(base, paths);
  }

  if (base.tag === "known" && base.sha !== pr.head.sha) {
    core.warning(
      `The checked out commit does not match the event PR's head. ${base.sha} != ${pr.head.sha}. Weird things may happen.`,
    );
  }

  // Write PR data for restyle to use. It expects a proper API type so we can't
  // do that with our fakePullRequest. This is also why centralize restyleArgs
  // handling here.
  const pullRequestJson = temp.path({ suffix: ".json" });
  fs.writeFileSync(pullRequestJson, JSON.stringify(pr));

  // Respect paths if given, otherwise pull PR changed files
  const restylePaths =
    paths.length === 0 ? await getPullRequestPaths(client, pr.number) : paths;

  const diff = await getPullRequestDiff(client, pr.number);

  return {
    number: pr.number,
    title: pr.title,
    headRef: pr.head.ref,
    headSha: pr.head.sha,
    restyleArgs: ["--pull-request-json", pullRequestJson].concat(restylePaths),
    restyleDiffBase: base,
    diff,
  };
}

function fakePullRequest(base: DiffBase, paths: string[]): PullRequest {
  return {
    number: 0,
    title: "Unknown",
    headRef: "unknown",
    headSha: "unknown",
    restyleArgs: paths,
    restyleDiffBase: base,
    diff: null,
  };
}

async function getPullRequestPaths(
  client: GitHubClient,
  number: number,
): Promise<string[]> {
  const files = await client.paginate(client.rest.pulls.listFiles, {
    ...github.context.repo,
    pull_number: number,
  });

  return files.map((f) => f.filename);
}

async function getPullRequestDiff(
  client: GitHubClient,
  number: number,
): Promise<string> {
  const response = await client.rest.pulls.get({
    ...github.context.repo,
    pull_number: number,
    mediaType: {
      format: "patch",
    },
  });

  // Custom mediatype is not expected by octokit types
  return response.data as unknown as string;
}

async function getDiffBase(): Promise<DiffBase> {
  try {
    const sha = await readProcess("git", ["rev-parse", "HEAD"]);
    return { tag: "known", sha };
  } catch {
    return { tag: "unknown" };
  }
}
