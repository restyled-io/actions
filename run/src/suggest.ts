/* Copyright (C) 2025 Patrick Brisbin
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
import { NonEmpty, nonEmpty } from "./non-empty";
import * as NE from "./non-empty";
import { PatchLine, parseGitPatches, parseGitPatch } from "./parse-git-patch";
import { PullRequestFile } from "./pull-request";

export type Suggestion = {
  path: string;
  description: string;
  code: string[];
  startLine: number;
  endLine: number;
  skipReason?: string;
};

export function suggest(
  baseFiles: PullRequestFile[],
  resolved: Suggestion[],
  patch: string,
): Suggestion[] {
  const patches = parseGitPatches(patch);
  const suggestions: Suggestion[] = [];

  patches.forEach((patch) => {
    patch.files.forEach((file) => {
      const groups = NE.groupBy(file.modifiedLines, (a, b) => {
        return a.tag === b.tag || (a.tag === "removed" && b.tag === "added");
      });

      groups.forEach((group) => {
        const removed = getRemoveLineNumbers(group);

        if (!removed) {
          return;
        }

        const suggestion: Suggestion = {
          path: file.afterName,
          description: (patch.message ?? "").replace(/\[PATCH.*] /, ""),
          code: getAddedLines(group),
          startLine: NE.head(removed),
          endLine: NE.last(removed),
        };

        if (!isOnAddedLines(baseFiles, suggestion)) {
          suggestion.skipReason = "suggestions can only be made on added lines";
        }

        if (resolved.some((r) => isSameLocation(r, suggestion))) {
          suggestion.skipReason = "previously marked resolved";
        }

        suggestions.push(suggestion);
      });
    });
  });

  return suggestions;
}

function getRemoveLineNumbers(
  lines: NonEmpty<PatchLine>,
): NonEmpty<number> | null {
  const acc: number[] = [];

  NE.toList(lines).forEach((line) => {
    if (line.tag === "removed") {
      acc.push(line.removedLineNumber);
    }
  });

  return nonEmpty(acc);
}

function getAddedLines(lines: NonEmpty<PatchLine>): string[] {
  const acc: string[] = [];

  NE.toList(lines).forEach((line) => {
    if (line.tag === "added") {
      acc.push(line.line);
    }
  });

  return acc;
}

function isOnAddedLines(
  baseFiles: PullRequestFile[],
  suggestion: Suggestion,
): boolean {
  const matched: PatchLine[] = [];
  const { path, startLine, endLine } = suggestion;
  const suggestionSize = endLine - startLine + 1;
  const lines = getPullRequestDiff(baseFiles, path);

  if (lines.length >= suggestionSize) {
    for (let i = startLine; i <= endLine; i++) {
      const line = lines.find((line) => {
        return line.tag === "added" && line.addedLineNumber === i;
      });

      if (line) {
        matched.push(line);
      }
    }
  }

  return matched.length == suggestionSize;
}

function getPullRequestDiff(
  files: PullRequestFile[],
  name: string,
): PatchLine[] {
  const file = files.find((f) => f.filename === name);

  if (!file) {
    // File not in PR
    return [];
  }

  const diff = [
    `diff --git a/${file.filename} b/${file.filename}`,
    "index 000000000..000000000 100644",
    `--- a/${file.filename}`,
    `+++ b/${file.filename}`,
    file.patch ?? "",
  ].join("\n");

  const patch = parseGitPatch(diff);
  const patchFile = (patch?.files ?? [])[0];

  if (!patchFile) {
    return [];
  }

  return patchFile.modifiedLines;
}

function isSameLocation(a: Suggestion, b: Suggestion): boolean {
  return (
    a.path === b.path && a.startLine == b.startLine && a.endLine == b.endLine
  );
}
