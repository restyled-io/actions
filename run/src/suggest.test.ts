/* Copyright (C) 2025 Patrick Brisbin
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
import * as fs from "fs";

import { PullRequestFile } from "./pull-request";
import { suggest } from "./suggest";

describe("suggest", () => {
  const cases = ["test-files/suggest/restyled-io-actions-110"];

  test.each(cases)("%s", (path) => {
    const baseFilesRaw = fs.readFileSync(`${path}/files.json`).toString();
    const baseFiles: PullRequestFile[] = JSON.parse(baseFilesRaw);
    const patch = fs.readFileSync(`${path}/restyled.patch`).toString();

    const suggestions = suggest(baseFiles, [], patch);

    expect(suggestions).toMatchSnapshot();
  });
});
