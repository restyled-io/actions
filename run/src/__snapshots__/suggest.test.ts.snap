// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`suggest suggest-test-files/restyled-io-actions-110 1`] = `
Array [
  Object {
    "code": Array [
      "| 492 | 494 |",
    ],
    "description": "Restyled by whitespace",
    "endLine": 51,
    "path": "run/src/parse-git-patch.test.ts",
    "startLine": 51,
  },
  Object {
    "code": Array [
      "|   3 |   3 |",
    ],
    "description": "Restyled by whitespace",
    "endLine": 59,
    "path": "run/src/parse-git-patch.test.ts",
    "startLine": 59,
  },
  Object {
    "code": Array [
      "|   8 |   9 |",
    ],
    "description": "Restyled by whitespace",
    "endLine": 65,
    "path": "run/src/parse-git-patch.test.ts",
    "startLine": 65,
  },
  Object {
    "code": Array [
      "|  21 |  24 |",
    ],
    "description": "Restyled by whitespace",
    "endLine": 77,
    "path": "run/src/parse-git-patch.test.ts",
    "startLine": 77,
  },
  Object {
    "code": Array [
      "|  21 |  22 |",
    ],
    "description": "Restyled by whitespace",
    "endLine": 90,
    "path": "run/src/parse-git-patch.test.ts",
    "startLine": 90,
  },
  Object {
    "code": Array [
      "| 111 | 112 |",
    ],
    "description": "Restyled by whitespace",
    "endLine": 102,
    "path": "run/src/parse-git-patch.test.ts",
    "startLine": 102,
  },
  Object {
    "code": Array [
      "|  21 |  22 |",
    ],
    "description": "Restyled by whitespace",
    "endLine": 105,
    "path": "run/src/parse-git-patch.test.ts",
    "startLine": 105,
  },
  Object {
    "code": Array [
      "| 111 | 112 |",
    ],
    "description": "Restyled by whitespace",
    "endLine": 372,
    "path": "run/src/parse-git-patch.test.ts",
    "startLine": 372,
  },
  Object {
    "code": Array [
      "|  42 |  43 |",
    ],
    "description": "Restyled by whitespace",
    "endLine": 413,
    "path": "run/src/parse-git-patch.test.ts",
    "startLine": 413,
  },
  Object {
    "code": Array [
      "| 25 | 25 |",
    ],
    "description": "Restyled by whitespace",
    "endLine": 504,
    "path": "run/src/parse-git-patch.test.ts",
    "startLine": 504,
  },
  Object {
    "code": Array [
      "| 30 | 29 |",
    ],
    "description": "Restyled by whitespace",
    "endLine": 510,
    "path": "run/src/parse-git-patch.test.ts",
    "startLine": 510,
  },
]
`;
